libraries/splunk_app.rb

```diff
@@ -126,13 +174,80 @@ def create_splunk_directory(path)
         dir.group(node['splunk']['group'])
         dir.mode('0755')
         dir.run_action(:create)
-        new_resource.updated_by_last_action(dir.updated_by_last_action?)
+        new_resource.updated_by_last_action(true) if dir.updated_by_last_action?
+      end
+
+      def download_and_install # rubocop:disable PerceivedComplexity, CyclomaticComplexity
+        return if new_resource.url.nil? || new_resource.url.empty?
+
+        expected_version = CernerSplunk::AppVersion.new new_resource.version
+        installed_version = CernerSplunk::AppVersion.new @current_resource.version
+
+        return if expected_version.type && expected_version == installed_version
+        Chef::Log.warn "Expected version not specified for #{new_resource.app}." unless expected_version.type
+        fail "Expecting to install prerelease on top of same released version for #{new_resource.app}" if expected_version.type == :prerelease && installed_version.type == :base && expected_version.base == installed_version.base
+
+        # In whyrun mode, the rest is not safe.
+        return if whyrun_mode?
+
+        filename = "#{Chef::Config[:file_cache_path]}/#{new_resource.app}.tgz"
+
+        download = Chef::Resource::RemoteFile.new(filename, run_context)
+        download.source(new_resource.url)
+        download.run_action(:create)
+        download.backup(false)
+
+        install_from_tar filename, expected_version, installed_version
+      ensure
+        download.run_action(:delete) if download
+      end
+
+      def install_from_tar(filename, expected_version, installed_version) # rubocop:disable PerceivedComplexity, CyclomaticComplexity
+        tarfile = CernerSplunk::TarBall.new(filename, prefix: new_resource.app, user: node['splunk']['user'], group: node['splunk']['group'])
+
+        fail "Downloaded tarball from '#{new_resource.url}' does not contain an app named '#{new_resource.app}'" if tarfile.num_files == 0
+
+        fail "Downloaded tarball for '#{new_resource.app}' has local entries" unless tarfile.count { |p, _| p.match %r{^[^/]+/local/.+} } == 0
+
+        app_conf = CernerSplunk::Conf.parse_string tarfile.get_file('default/app.conf')
+        tar_version = CernerSplunk::AppVersion.new((app_conf['launcher'] || {})['version'])
+
+        fail "Bad Developer!!!! The downloaded tar for #{new_resource.app} does not contain a version in app.conf!" if tar_version.type.nil?
+
+        if expected_version.type && tar_version != expected_version
+          fail "Expected version #{expected_version} does not match tar version #{tar_version} for #{new_resource.app}" unless expected_version.type == :base && tar_version.base == expected_version.base
+        end
+
+        return if tar_version == installed_version
```

Comment:
I noticed in your test you have my_app as [1.0 SNAPSHOT in the app.conf](https://github.com/cerner/cerner_splunk/pull/48/files#diff-9cee7098655ddcc82045b71f38745918R2) but then [replace "SNAPSHOT" with a timestamp during provisioning of the VM](https://github.com/cerner/cerner_splunk/pull/48/files#diff-23b6f443c01ea2efcb4f36eedfea9089R156).  I guess if you didn't do something like that then this line would prevent newer versions of the SNAPSHOT from being installed on subsequent chef-client runs with a newer 1.0-SNAPSHOT tar, right?

My team has an app tarball but currently our snapshot tarball builds would always have something like "1.0 SNAPSHOT" in the app.conf.  Actually, we currently have "1.0-SNAPSHOT" but reviewing this I noticed that is a violation of the app.conf spec as it dictates a space between the base an pre-release version portions.  Anyway, it seems we will need to do some replacement of "SNAPSHOT" with a timestamp or build number for this to work for us.  That should be ok, but one other thing we are doing is setting the build number in the app.conf and we are setting that differently on every build.  An alternative would be for you to look at that as well.  Then perhaps you would re-deploy if the build numbers differ even though the versions match.  I do sort of like putting a build identifier in the pre-release portion of the version itself though as it makes the exact SNAPSHOT version visible in installed apps list in the Web UI.

I guess the question is what are your thoughts about the build attribute?  The spec suggests it is required and you have left it off you my_app app.conf example BTW.